### PR TITLE
[WIP] Handling Pipeline as stream

### PR DESCRIPTION
**Pull request by tpluscode**
My first attempt at working directly with streams instead of the pipeline object.

I'm not super satisfied by the changes so far:

The `Pipeline` class is no longer a stream itself. Instead of that I added the async `getStream` method. That however makes the usage a tad more awkward. And the pipeline itself is still needed to pass around so that we can access the context, variables, etc.

### CODE REVIEW

File: lib/loader/pipeline.js [hunk 1]
```diff
@@ -5,15 +5,13 @@ const Pipeline = require('../pipeline')
 function loader (term, dataset, { context, variables, basePath, loaderRegistry }) {
   const node = cf(dataset).node(term)
 
-  const type = node.out(ns.rdf('type'))
-
   const pipelineInit = { basePath, context, variables, loaderRegistry }
 
-  if (type.term && type.term.equals(ns.p('Pipeline'))) {
+  if (node.has(ns.rdf('type'), ns.p('Pipeline')).values.length > 0) {
```
Comment (tpluscode) [inline lib/loader/pipeline.js:10]: Can this be simpler?

File: examples/forEach.ttl [hunk 2]
```diff
@@ -24,6 +24,6 @@
 <http://example.org/pipeline#serializeJson> code:implementedBy [ a code:EcmaScript ;
             code:link <file:steps/serializeJson.js> ] .
 
-<http://example.org/pipeline#subPipeline> a p:Pipeline ;
+<http://example.org/pipeline#subPipeline> a p:Pipeline, p:WritablePipeline ;
```
Comment (tpluscode) [inline examples/forEach.ttl:27]: Not happiest about this, but pipeline would have to be marked as writable explicitly

File: lib/loader/registry.js [hunk 3]
```diff
@@ -36,7 +36,14 @@ class LoaderRegistry {
...
       loader = this._literalLoaders.get(node.term.datatype.value)
     } else {
-      loader = this._nodeLoaders.get(node.out(ns.rdf('type')).value)
+      const types = node.out(ns.rdf('type')).terms
+
+      for (let i = 0; i < types.length; i++) {
+        loader = this._nodeLoaders.get(types[i].value)
+        if (loader) {
+          break
+        }
```
Comment (tpluscode) [inline lib/loader/registry.js:43]: I feel like I'm not thinking straight any more. How do I make this simpler?

Comment (bergos) [reply lib/loader/registry.js:43]: Maybe a `for of` loop on `node.out(ns.rdf('type')).values` could make it simpler.

**Comment by bergos** [comment]
I think it will not work with duplexify or just returning the first or last stream. With the async init step, it must be integrated to the `_write` or `_read` method, like it's already done (master + PR). A async factory could be used, but for consistent logging it's nicer if it's done in the stream.

But to cover the different cases, more classes are required. I would propose that the `Pipeline` class doesn't inherit from any class, but implements the logic for `Readable` and `Writeable` (`_write`, `_read`). `ReadablePipleine`, `WritablePipeline` and `DuplexPipeline` inherit from the matching stream class. The classes will create their own pipeline object instances and just forward the calls to the matching methods in the wrapped pipeline:

```
ReadablePipeline extends Readable {
  constructor() {
    this.pipeline = new Pipeline({ readable: true, stream: this })
  }

  _read () {
    return this.pipeline._read()
  }
}
```

That should avoid code duplication. Only the loader has to figure out which constructor to call. Also flags for read/write mode should be forwarded to the pipeline object for validation the first/last step. Forwarding the stream instance to the wrapped pipeline in the constructor would allow the pipeline code to decided which methods to attach (e.g. `stream.clone = this.clone.bind(this)`)

**Comment by tpluscode** [comment]
> I think it will not work with duplexify or just returning the first or last stream

That is not the case anymore. The code no longer uses duplexity but instead forwards writes from the pipeline to its streams. The rest is largely unchanged.

**Comment by bergos** [comment]
implemented in #31

**Pull request closed without merging**